### PR TITLE
build(betas): fastlane + ci setup

### PR DESCRIPTION
**Pull request by rajsam003**
This PR resolves [MOPLAT-277]

### Description

Adds beta distribution via fastlane for iOS, circle ci setup for creating nightly betas and running tests and checks on PRs.
Mostly copied from Eigen with some tweaks and removals for Folio. 

- Adds a fastfile with beta distribution, release checks, taken from Eigen
- Adds a circle config for running tests, building betas, checking PRs 
- Adds missing icon assets required for beta submission
- Adds a danger config copied from Eigen
- Adds secrets baseline to allow check in PRs

#### Follow-ups

- [ ] Android beta distribution and store setup [TICKET]
- [ ] Sentry integration and project setup [TICKET]
- [ ] Fix or ignore typescript and lint failures and enable those checks [TICKET]
- [ ] Finish detect secrets integration, setup pre-commit hooks [TICKET]
- [ ] Update metadata and screenshots provided with app submission [TICKET]
- [ ] Enable branch protection for main branch

#### Nice to do

- [ ] Separate out app specific config to a centralized place so it is easier to setup ci for new projects
- [ ] Review our ci scripts, organize and remove redundancies, simplify if possible, maybe package ?

### PR Checklist

- [ ] I tested my changes on **iOS** and **Android**.
- [ ] I added screenshots or videos to illustrate my changes.
- [ ] I added Tests and Stories for my changes.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to run this PR on the simulator or device.

Need help with something? Have a look at our [docs], or get in touch with us.

[docs]: /docs/README.md

[MOPLAT-277]: https://artsyproduct.atlassian.net/browse/MOPLAT-277?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  `node` failed.
  </td>
    </tr>
  </tbody>
</table>



### Log


```sh
Error:  Error: Could not find a 'dangerfile.js' or 'dangerfile.ts' in the current working directory (/home/circleci/project).
    at dangerfilePath (/home/circleci/project/node_modules/danger/distribution/commands/utils/fileUtils.js:28:11)
    at /home/circleci/project/node_modules/danger/distribution/commands/danger-runner.js:90:61
    at step (/home/circleci/project/node_modules/danger/distribution/commands/danger-runner.js:34:23)
    at Object.next (/home/circleci/project/node_modules/danger/distribution/commands/danger-runner.js:15:53)
    at fulfilled (/home/circleci/project/node_modules/danger/distribution/commands/danger-runner.js:6:58)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)

```

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against a87e5dcdc8f59c19dda99c349cc8042adcafb4ca
</p>

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,26 +1,514 @@
...
+    steps:
+      - restore_cache:
+          keys:
+            - v1-jest-{{ .Branch }}
+      # - run:
+      #     name: Run TypeScript check
+      #     command: yarn type-check
+      # - run:
+      #     name: Run TSLint check
+      #     command: yarn lint
```
Comment (brainbicycle) [inline .circleci/config.yml:87]: these are failing for me with many failures, I will create a ticket to fix and enable

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,26 +1,514 @@
...
+    macos:
+      xcode: 13.2.1
+
+    steps:
+      - checkout
+      - generate-checksums
+      - switch-ruby
+      - install-gems
+      - run:
+          name: Check flags
```
Comment (brainbicycle) [inline .circleci/config.yml:238]:
```suggestion
          name: Check if app has been approved for release
```
think this is wrong in Eigen as well

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,26 +1,514 @@
...
+  await-previous-builds:
+    parameters:
+      branch:
+        description: The branch to wait on
+        type: string
+    steps:
+      - queue/until_front_of_line:
+          time: "30" # how long a queue will wait until the job exits
+          only-on-branch: <<parameters.branch>> # restrict queueing to a specific branch (default *)
+  switch-ruby:
```
Comment (brainbicycle) [inline .circleci/config.yml:19]: think this is actually unnecessary if you use a supported ruby and have a version file, try removing

File: dangerfile.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    )
+  }
+}
+;(async function () {
+  const newCreatedFileNames = getCreatedFileNames(danger.git.created_files)
+
+  preventUsingMoment()
+  preventUsingTestRenderer()
+  verifyRemainingDevWork()
+  useWebPs(newCreatedFileNames)
```
Comment (brainbicycle) [inline dangerfile.ts:116]: these were taken from Eigen, do we want all of these?

File: fastlane/Fastfile [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+  end
+
+  root = File.expand_path('..', __dir__)
+  bundle_version = `/usr/libexec/PlistBuddy -c "print CFBundleVersion" #{File.join(root, 'ios/Energy/Info.plist')}`.strip
+
+  tag_and_push(tag: "ios-#{latest_version}-#{bundle_version}")
+
+  # important! this much match the release version specified
+  # in Eigen in order for sourcemaps to work correctly
+  # sentry_release_name = "ios-#{latest_version}-#{bundle_version}"
```
Comment (brainbicycle) [inline fastlane/Fastfile:66]: commented out until we have sentry project setup and integrated

File: ios/Energy/Images.xcassets/AppIcon.appiconset/Contents.json [hunk 6]
```diff
@@ -1,61 +1,115 @@
 {
   "images" : [
     {
-      "filename" : "artsy-icon-blue 1-20@2x.png",
+      "filename" : "ArtsyIcon-iPhoneNotification-20x20@2x.png",
```
Comment (brainbicycle) [inline ios/Energy/Images.xcassets/AppIcon.appiconset/Contents.json:4]: standardized names and added missing assets required for beta release

File: scripts/distribute-ios [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+set -euxo pipefail
+
+
+./scripts/setup-distribute
+yarn relay
```
Comment (brainbicycle) [inline scripts/distribute-ios:6]: think this is unnecessary, try removing
```suggestion
```

File: .circleci/config.yml [hunk 8]
```diff
@@ -1,26 +1,514 @@
...
+          paths:
+            - .manifests
+            - dist
+
+  build-test-app-ios:
+    environment:
+      BUNDLE_PATH: .vendor # path to install gems and use for caching
+
+    macos:
+      xcode: 13.2.1
```
Comment (brainbicycle) [inline .circleci/config.yml:280]: can probably use the latest Xcode, tried this when coming across an archiving issue that I believe was unrelated, try again with latest

File: .tool-versions [hunk 9]
```diff
@@ -1 +1,2 @@
-ruby 2.7.4
+ruby 2.7.6
+nodejs 16.6.0
```
Comment (brainbicycle) [inline .tool-versions:2]: I tried with node 18 and ran into this issue when trying to archive on CI: https://github.com/webpack/webpack/issues/14532
If this interferes with anything let me know

File: fastlane/Fastfile [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+
+date_str = DateTime.now.strftime("%Y.%m.%d.%H")
+lane :set_build_version_ios do
+  ios_targets.each do |id, meta|
+    system("/usr/libexec/PlistBuddy -c \"Set CFBundleShortVersionString #{app_json['version']}\" #{meta[:plist_path]}")
+    system("/usr/libexec/PlistBuddy -c \"Set CFBundleVersion #{date_str}\" #{meta[:plist_path]}")
+  end
+end
+
+# lane :promote_beta_ios do
```
Comment (brainbicycle) [inline fastlane/Fastfile:114]: enable and update any eigen specific values

File: .tool-versions [hunk 11]
```diff
@@ -1 +1,2 @@
-ruby 2.7.4
+ruby 2.7.6
+nodejs 16.10.0
```
Comment (pvinis) [inline .tool-versions:2]:
```suggestion
nodejs 16.10.0
```

File: .bundle/config [hunk 12]
```diff
@@ -0,0 +1,2 @@
+BUNDLE_PATH: ".vendor"
+BUNDLE_FORCE_RUBY_PLATFORM: 1
```
Comment (pvinis) [inline .bundle/config:2]:
```suggestion
BUNDLE_FORCE_RUBY_PLATFORM: 1
```
newline

File: .circleci/config.yml [hunk 13]
```diff
@@ -1,26 +1,501 @@
...
+            - v1-pods-{{ checksum ".manifests/cocoapods" }}
+      - run:
+          name: Install Pods
+          command: cd ios; bundle exec pod check --ignore-dev-pods || bundle exec pod install; cd ..;
+      - save_cache:
+          key: v1-pods-{{ checksum ".manifests/cocoapods" }}
+          paths:
+            - ios/Pods
+            # this changes after pod install because we don't install dev tools on CI
+            # so we save it to the cache to avoid comparison errors with Pods/Manifest.lock
```
Comment (pvinis) [inline .circleci/config.yml:120]: nice

File: app.json [hunk 14]
```diff
@@ -1,4 +1,5 @@
 {
   "name": "Energy",
-  "displayName": "Energy"
+  "displayName": "Energy",
+  "version": "3.0.0"
```
Comment (pvinis) [inline app.json:4]: BOOM

File: fastlane/Pluginfile [hunk 15]
```diff
@@ -0,0 +1 @@
+gem 'fastlane-plugin-sentry'
```
Comment (pvinis) [inline fastlane/Pluginfile:1]:
```suggestion
gem 'fastlane-plugin-sentry'
```
newline

File: ios/Podfile [hunk 16]
```diff
@@ -44,5 +47,5 @@ target 'Energy' do
     end
 
   end
-  
+
 end
```
Comment (pvinis) [inline ios/Podfile:51]:
```suggestion
end
```
newline

File: package.json [hunk 17]
```diff
@@ -100,14 +101,17 @@
...
     "eslint": "^7.14.0",
     "eslint-config-prettier": "^8.3.0",
+    "eslint-plugin-prettier": "^4.2.1",
     "eslint-import-resolver-typescript": "^3.1.1",
     "eslint-plugin-import": "^2.26.0",
     "graphql": "^15.0.0",
     "husky": "^7.0.4",
     "jest": "^26.0.0",
     "jest-environment-jsdom": "^26.0.0",
+    "jest-junit": "^14.0.0",
```
Comment (pvinis) [inline package.json:114]: sweet 👌

File: dangerfile.ts (hunk 4 above)
Comment (pvinis) [reply dangerfile.ts:116]: lets keep them, and probably we can merge them or move them elsewhere at some point? maybe a nice little ci orb/script that will run for both eigen and energy, from some artsy/mobile-ci repo 🤔

**Review by pvinis — APPROVED** [review]
wow, great work!

**Comment by pvinis** [comment]
ill assign back to you, so you can merge whenever you think is best 👍

File: scripts/generate-manifest.js [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+  const outStream = fs.createWriteStream(outFileName, { encoding: "utf-8" })
+
+  for (const file of matchingFiles) {
+    outStream.write(`${file}\t${hash(file)}\n`)
+  }
+
+  outStream.close()
+}
+
+run()
```
[the remark this answers is not in the document]
Comment (brainbicycle) [reply scripts/generate-manifest.js:65]: we use them for caching on circle, the manifests are checked when determining whether to redownload or rebuild a step, not sure how well it is still working but that is the intent

**Pull request merged**